subsys/bluetooth/host/id.c

```diff
@@ -777,6 +795,25 @@ void bt_id_add(struct bt_keys *keys)
 		goto done;
 	}
 
+	/* Replace the old key, if it exist with the new since the IRK may be different
+	 * from the previous bonding.
+	 */
+	if (bt_keys_find_addr(keys->id, &keys->addr) != NULL) {
```

Comment:
shouldn't this be handled before check for RL size exceed above?

scenario for RL size==2:
RL: ID1 ID2
you want to add key with ID1 but RL is already 'full' so it switch to host resolving
ID2 is removed, and thus we got back to controller resolving
RL now contains old ID1 info...


TBH I start to wonder if we shouldn't just always clear RL and fully restore it when adding or removing IDs

Reply:
It looks like the RL in the controller is cleared when trying to add a new IA to the RL and it is full.

Reply:
So going back to controller resolving will not have ID1 info in it, right?

Reply:
I think the current implementation works for the described scenario since a BT_HCI_OP_LE_CLEAR_RL command is sent to the controller if the controller resolve list is full. Do you agree?

Reply:
Now that you mention invalid rl_entries. I can see that rl_entries > rl_size is used to mark that host side resolving is used, but then it does not count the number of entries anymore since they were cleared in the controller's resolve list. How does that work?

Also the hci_id_del() call I added (to overwrite the existing entry), should it be decremented if it succeeds? To keep rl_entries valid. Otherwise it seems to me like an overwrite would increment the number of entries, but it should remain the same. What do you think @sjanc ?

Reply:
I think this might resolve the issues you highlighted:

```c
	key_already_added	= (bt_keys_find_addr(keys->id, &keys->addr) != NULL);
	if ((bt_dev.le.rl_entries == bt_dev.le.rl_size) && (key_already_added == false)) {
		BT_WARN("Resolving list size exceeded. Switching to host.");

		err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_CLEAR_RL, NULL, NULL);
		if (err) {
			BT_ERR("Failed to clear resolution list");
			goto done;
		}

		bt_dev.le.rl_entries++;

		goto done;
	}

	/* Replace the old key, if it exist with the new since the IRK may be different
	 * from the previous bonding.
	 */
	if (key_already_added == true) {
		/* The identity address was found in the host's key pool.
		 * So it should also be present in the controller's resolve list.
		 * Delete the existing entry before adding the new in case
		 * the IRK is new.
		 */
		err = hci_id_del(&keys->addr);
		if (err) {
			BT_ERR("Failed to remove previous IRK in the controller");
			/* Key_pool had the identity address, but the controller did not.
			 * Continue and attempt to add the identity address to the
			 * controller.
			 */
		}
		else {
			/* hci_id_del succeeded. Decrement the number of entries */
			bt_dev.le.rl_entries--;
		}
	}

	err = hci_id_add(&keys->addr, keys->irk.val);
	if (err) {
		BT_ERR("Failed to add IRK to controller");
		goto done;
	}

	bt_dev.le.rl_entries++;

```
Even though I am still in doubt about rl_entries++ when switching to host side resolving, but it was there before.

I am currently testing this. But what do you think about it @sjanc ?

Reply:
as I already mentioned, ID code has become quite complicated :-)  my understanding is that bt_dev.le.rl_entries is number of entries that would have been added to RL (assuming infinite RL where host fallback is just implementation detail) and it feels that bt_id_add() was not suppose to be used to update RL entry.

so either bt_id_add() is extended with proper support for "update" scenario (maybe it would as simple as always call bt_id_remove() before calling bt_id_add() ? drawback would be case that could unnecessary toggle host<->ll resolving)

Reply:
yes, something like your snipped,  but I think you should check for key_already_added at the very begging,  ie to not increment rl_entries if host fallback is already enabled

Reply:
But if host side resolving is already enabled when bt_id_add() is called, then bt_id_add() will not be doing any hci_id_del or hci_id_add(). It will be the host code that handles overwriting the key, right? So we do not need to check if key_already_added is true in the first check for host side resolving. As I understand it. Do you agree @sjanc ?

Reply:
yes, but it will stil increment bt_dev.le.rl_entries

Reply:
But bt_id_add cannot know what the host side does. By incrementing bt_dev.le.rl_entries in the beginning of bt_id_add if host side resolving is enabled, then it assumes that host side is always adding. But that is a different issue, right?

Reply:
bt_id_add() is now unchanged.